### PR TITLE
[Merged by Bors] - feat: have `notation3` use elaborator when generating matchers, add support for pi/lambda

### PR DESCRIPTION
**Pull request by kmill**
`notation3` was generating matchers directly from syntax, which included a half-baked implementation of a term elaborator. This switches to elaborating the term and then generating matchers from the elaborated term. This
1. is more robust and consistent, since it uses the main elaborator and one can make use of other notations
2. has the nice side effect of adding term info to expansions in the `notation3` command
3. can unfortunately generate matchers that are more restrictive than before since they also match against elaborated features such as implicit arguments.

We now also generate matchers for expansions that have pi types and lambda expressions.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by kmill** [comment]
@digama0 Here's something you'd suggested during the review of the first implementation of the `notation3` delaborator. I've spot-checked that `notation3` notation still works with this, though there's one case involving `Funlike.coe` that generates a too-restrictive matcher (I've left a TODO).

File: Mathlib/Mathport/Notation.lean [hunk 1]
```diff
@@ -170,51 +170,99 @@ def matchFVar (userName : Name) : Matcher := fun s => do
...
-    let userName ← fvarId.getUserName
-    return (none, ← ``(matchFVar $(quote userName)), ← collectIdxs (← inferType fe))
+/-- Fully elaborates the term `patt`, allowing typeclass inference failure.
+Instantiates all assigned metavariables. -/
+def elabPattern (patt : Term) (expectedType? : Option Expr) : TermElabM Expr := do
+  withTheReader Term.Context (fun ctx => { ctx with ignoreTCFailures := true }) <|
+    Term.withSynthesizeLight do
+      let t ← Term.elabTerm patt expectedType?
+      Term.synthesizeSyntheticMVars (mayPostpone := false) (ignoreStuckTC := true)
+      instantiateMVars t
```
Comment (kmill) [inline Mathlib/Mathport/Notation.lean:212]: Should we put this in the Util folder? If so, perhaps `Util/Elab.lean`?

File: Mathlib/Mathport/Notation.lean [hunk 2]
```diff
@@ -158,63 +160,114 @@ def matchVar (c : Name) : Matcher := fun s => do
...
+  for name in boundNames do
+    let fvarId ← mkFreshFVarId
+    lctx := lctx.mkLocalDecl fvarId name (← mkFreshTypeMVar)
+    boundFVars := boundFVars.insert fvarId name
+  return (lctx, boundFVars)
+
+/-- Given an expression, generate a matcher for it.
+The `boundFVars` hash map records which state variables certain fvars correspond to.
+The `localFVars` hash map records which local variable the matcher should use for an exact
+expression match. -/
```
Comment (eric-wieser) [inline Mathlib/Mathport/Notation.lean:222]: Can you explain the return value here?

Comment (kmill) [inline Mathlib/Mathport/Notation.lean:222]:
```suggestion
expression match.

If it succeeds generating a matcher, returns
1. a list of keys that should be used for the `delab` attribute
   when defining the elaborator
2. a `Term` that represents a `Matcher` for the given expression `e`. -/
```

Comment (kmill) [reply Mathlib/Mathport/Notation.lean]: Done.

**Comment by PatrickMassot** [comment]
Thanks a lot for all your work on this!
bors merge

**Pull request closed without merging**